### PR TITLE
Dev server should not exit when it receives EOF (-1)

### LINKED ISSUE #11086: When running Play from sbt when piping input/output, web-application shuts down immediately after starting
I am using _Play_ with _IntelliJ IDEA 2021.3.1 (Ultimate Edition)_, together with the _IntelliJ Play Framework_ plugin (v 213.5744.125). I am reporting this problem here, because I believe the issue lies with how the _Play_ web-server executes, rather than with how _IntelliJ_ runs the _Play_ application. The application is using _SBT_ v1.5.8 to build, with _Scala_ 2.13.7.

The problem is that the _Play_ web server immediately shuts down after starting (via either the _IntelliJ Play Framework_ plugin's _Play 2 App_ or  `sbt run` run configurations), while piping it's terminal input/output to _IntelliJ_. This did not happen with the previous Play version that I was using (v2.8.8). Furthermore, it's still possible to run `sbt run` from the command line, and have the web application start successfully. However, I could be wrong, and it's possible this is a latent issue with _IntelliJ IDEA_.

I suspect that this may be a symptom of the recent fix to #11032, and the changes made between release 2.8.8 and 2.8.11 that caused that issue to arise.

### Play Version
2.8.12

### API

Scala

### Operating System

Windows 10 Pro 64-bit (21H1)

### JDK

openjdk version "11.0.13" 2021-10-19 LTS
OpenJDK Runtime Environment Zulu11.52+13-CA (build 11.0.13+8-LTS)
OpenJDK 64-Bit Server VM Zulu11.52+13-CA (build 11.0.13+8-LTS, mixed mode)

### Expected Behavior

1. Start web-application from within _IntelliJ_ using _IntelliJ Play Framework_'s _Play 2 App_ run configuration, or a regular _IntelliJ_ _sbt run task_ run configuration.

### Actual Behavior

1. Web application starts, then immediately terminates:

Output from a "Play 2 App" run configuration:

```plain
"C:\Program Files\Zulu\jdk-11\bin\java.exe" -Dfile.encoding=UTF8 -Djline.terminal=none -Dsbt.global.base=C:\Users\Me\AppData\Local\Temp\sbt-global-pluginstub -Dsbt.log.noformat=true -XX:+CMSClassUnloadingEnabled -classpath C:\Users\Me\AppData\Roaming\JetBrains\IntelliJIdea2021.3\plugins\Scala\launcher\sbt-launch.jar xsbt.boot.Boot "project root" run
[info] welcome to sbt 1.5.8 (Azul Systems, Inc. Java 11.0.13)
[info] loading settings for project mywebapp-build from CoffeeScript.sbt,Play.sbt,Release.sbt ...
[info] loading project definition from C:\Users\Me\src\MyWebApp\project
[info] loading settings for project root from build.sbt,version.sbt ...
[info]   __              __
[info]   \ \     ____   / /____ _ __  __
[info]    \ \   / __ \ / // __ `// / / /
[info]    / /  / /_/ // // /_/ // /_/ /
[info]   /_/  / .___//_/ \__,_/ \__, /
[info]       /_/               /____/
[info] 
[info] Version 2.8.12 running Java 11.0.13
[info] 
[info] Play is run entirely by the community. If you want to keep using it please consider donating:
[info] https://www.playframework.com/sponsors
[info] 
[info]   __              __
[info]   \ \     ____   / /____ _ __  __
[info]    \ \   / __ \ / // __ `// / / /
[info]    / /  / /_/ // // /_/ // /_/ /
[info]   /_/  / .___//_/ \__,_/ \__, /
[info]       /_/               /____/
[info] 
[info] Version 2.8.12 running Java 11.0.13
[info] 
[info] Play is run entirely by the community. If you want to keep using it please consider donating:
[info] https://www.playframework.com/sponsors
[info] 

--- (Running the application, auto-reloading is enabled) ---

[info] p.c.s.AkkaHttpServer - Listening for HTTP on /127.0.0.1:9000

Server started, use Alt+D to stop

[info] p.a.i.l.c.CoordinatedShutdownSupport - Starting synchronous coordinated shutdown with ServerStoppedReason reason and 35000 milliseconds timeout
[info] p.c.s.AkkaHttpServer - Stopping Akka HTTP server...
[info] p.c.s.AkkaHttpServer - Terminating server binding for /127.0.0.1:9000
[info] p.c.s.AkkaHttpServer - Running provided shutdown stop hooks

[success] Total time: 3 s, completed Jan 6, 2022, 4:04:42 PM

Process finished with exit code 0
```

Output from a `sbt run` _sbt Task_ run configuration:
```plain
"C:\Program Files\Zulu\jdk-11\bin\java.exe" -XX:+CMSClassUnloadingEnabled "-javaagent:C:\Program Files\JetBrains\IDEAU\lib\idea_rt.jar=53264:C:\Program Files\JetBrains\IDEAU\bin" -Dfile.encoding=windows-1252 -classpath C:\Users\Me\AppData\Roaming\JetBrains\IntelliJIdea2021.3\plugins\Scala\launcher\sbt-launch.jar xsbt.boot.Boot run
[info] welcome to sbt 1.5.8 (Azul Systems, Inc. Java 11.0.13)
[info] loading global plugins from C:\Users\Me\.sbt\1.0\plugins
[info] loading settings for project mywebapp-build from CoffeeScript.sbt,Play.sbt,Release.sbt ...
[info] loading project definition from C:\Users\Me\src\MyWebApp\project
[info] loading settings for project root from build.sbt,version.sbt ...
[info]   __              __
[info]   \ \     ____   / /____ _ __  __
[info]    \ \   / __ \ / // __ `// / / /
[info]    / /  / /_/ // // /_/ // /_/ /
[info]   /_/  / .___//_/ \__,_/ \__, /
[info]       /_/               /____/
[info] 
[info] Version 2.8.12 running Java 11.0.13
[info] 
[info] Play is run entirely by the community. If you want to keep using it please consider donating:
[info] https://www.playframework.com/sponsors
[info] 

--- (Running the application, auto-reloading is enabled) ---

[info] p.c.s.AkkaHttpServer - Listening for HTTP on /127.0.0.1:9000

(Server started, use Enter to stop and go back to the console...)

[info] p.a.i.l.c.CoordinatedShutdownSupport - Starting synchronous coordinated shutdown with ServerStoppedReason reason and 35000 milliseconds timeout
[info] p.c.s.AkkaHttpServer - Stopping Akka HTTP server...
[info] p.c.s.AkkaHttpServer - Terminating server binding for /127.0.0.1:9000
[info] p.c.s.AkkaHttpServer - Running provided shutdown stop hooks

[success] Total time: 5 s, completed Jan 6, 2022, 4:07:38 PM

Process finished with exit code 0
```

### PR DESCRIPTION
**Pull request by mkurz**
Fixes #11086

**TLDR:** After quite some testing I am almost 100% sure we don't need to quit the `sbt run` task on `EOF` (-1). `EOF` will never be received when using `CTRL+D` or the enter key (at least when using Linux/bash)

@MikeEngUSA @gmethvin Can you please test? I published a snapshot that is based on Play 2.8.12 with this PR applied. The version is `2.8.12+1-f35a1c22-SNAPSHOT`, you also have to add `resolvers += Resolver.sonatypeRepo("snapshots")` to both your `project/plugins.sbt` and `build.sbt`.
This is the project I used for testing: https://github.com/mkurz/play-eof-test/blob/master/project/plugins.sbt As you can see I also published a second snapshot (`2.8.12+2-f95484fb-SNAPSHOT`) which also [prints out the character that was received](https://github.com/mkurz/playframework/commit/f95484fb3c7550dbceb3e88bf287f217bea0cd07). You can use that as well if you want to get more insights. It could happen that when you test piping you get floated with `-1`...
If you both test on MacOS and Windows with [a few sbt versions](https://github.com/mkurz/play-eof-test/blob/master/project/build.properties) (not need to test all, maybe 1.6.1, 1.3.13 and 0.13.18?) by running `sbt run` as well as `sbt run > console.log 2>&1` and maybe @MikeEngUSA the various IntelliJ run configs, and everything is OK I will merge this and cut Play 2.8.13. Thanks!

---

Long story:

In #10890 we switched to `System.in`, which, starting with sbt 1.4+, is now non-blocking (that's why #11061 was required as well).

Now, when using `System.in`, what's happening on my Ubuntu machine and obviously also on Windows, is that when piping `sbt`s input/output, Play (jline?) receives `EOF` (-1) for some reason. This wasn't the case when using `FileDescriptor.in`. (Or maybe that's because `System.in` is non-blocking in sbt 1.4+ and `FileDescriptor.in` isn't? To be honest I don't know exactly).

For some reason, when switching to sbt in Play 2.2, the `run` task should also exit when it receives `EOF`/`-1`.
Here is the commit that introduced that: https://github.com/playframework/playframework/commit/bfbc16586d1307ddae51b5bf4f886ec6e4bb14da#diff-e5676a6f7f1849c865bd5e36be5293eea204aed23eec93a5421f39f14116fff8R31
The comment "_Note: we have to listen to -1 for jline2, for some reason..._" indicates that there was something fishy going on back then, maybe a bug in jline that made it necessary to also treat -1 as `CTRL-D` in some environments? I don't know. Because neither the comment nor the pull request this commit was included in didn't go into more details why catching `-1` is necessary, I had a look if I can find a jline bug that could have caused problems back then: So maybe https://github.com/jline/jline2/commit/39bbc9208f0a50bc20f8214a4f4f1005297d6850 or https://github.com/jline/jline2/commit/cae777dfeaa62bfd1aa362acad4b97decf6c91da could have been the reason to also listen for `-1`. However I don't know and will probably never know. There is also [this thread](https://github.com/jline/jline2/pull/67) where someone says `ctrl-D` sends `null` (instead of 4), but then some says `"...both Ctrl+C and Ctrl+D are read as litteral 3 and 4..."` so it seems there was at least something going on back then.

What's interesting as well is, that _before_ the commit I referenced above (which made it into Play 2.2.0-M3), the `waitForKey` was _not_ listening for `-1`: https://github.com/playframework/playframework/blob/2.2.0-M2/framework/src/sbt-plugin/src/main/scala/PlayRun.scala#L24 (this is from the `2.2.0-M2` tag)

Anyway.
Because it is not clear to me why `EOF` should be treated like `CTRL-D` (4), I did quite some testing to find out if there is a scenario that causes `EOF` (-1) to be received.  And I could **not** find one, at least not when using Ubuntu.
To test, I just printed the detected char out on the console, see https://github.com/mkurz/playframework/commit/f95484fb3c7550dbceb3e88bf287f217bea0cd07
Testing in with sbt `0.13.18`, `1.3.13`, `1.4.9`, `1.5.8` and `1.6.1` with `sbt run` it alway correctly printed `4` for `CTRL+D` and `13` for the enter key. `-1` was never seen. Only when piping via `sbt run > console.log 2>&1` or using Intellij's `Play 2 App` the output gets floated with `-1` (that's why when reacting to `-1` previously to this PR has stopped the run task immediately)

### CODE REVIEW

**Comment by mkurz** [comment]
@Mergifyio backport master 2.7.x

**Comment by gmethvin** [comment]
@mkurz Everything still works fine on Mac.

I tested on various sbt versions with your version that prints the character. On sbt 1.4.1 and later, I get `Character detected: -1` repeating seemingly indefinitely. On sbt versions before that I don't get any output. I suspect it was due to this change in sbt: https://github.com/sbt/sbt/pull/5975.

**Review by gmethvin — APPROVED** [review]

**Comment by ndeverge** [comment]
Hello,

I'm on MacOS, and I had the same issue with IntelliJ.
I tested the snapshot version `2.8.12+2-f95484fb-SNAPSHOT` on my own project, it displays "Character detected: -1" repeatedly.

With version `2.8.12+1-f35a1c22-SNAPSHOT`, it works correctly with sbt 1.5.5, 1.6.1 and 1.3.13 (can't test for 0.13.18 as my project is not compatible with it).

I can't test this on Windows.

**Comment by MikeEngUSA** [comment]
@mkurz OK, I've tested this fix, and the web-application no longer shuts down when running from inside _IntelliJ_. However, it does produce a lot of `Character detected: -1` messages (1,903 of them in total), before moving on to compile the web-application to satisfy a page request.

Here's the output:
```plain
[info] [launcher] getting org.scala-sbt sbt 1.5.8  (this may take some time)...
[info] [launcher] getting Scala 2.12.14 (for sbt)...
[info] welcome to sbt 1.5.8 (Azul Systems, Inc. Java 11.0.13)
[info] loading settings for project mywebapp-build from CoffeeScript.sbt,Play.sbt,Release.sbt ...
[info] loading project definition from C:\Users\Me\src\MyWebApp\project
[info] loading settings for project root from build.sbt,version.sbt ...
[info]   __              __
[info]   \ \     ____   / /____ _ __  __
[info]    \ \   / __ \ / // __ `// / / /
[info]    / /  / /_/ // // /_/ // /_/ /
[info]   /_/  / .___//_/ \__,_/ \__, /
[info]       /_/               /____/
[info] 
[info] Version 2.8.12+2-f95484fb-SNAPSHOT running Java 11.0.13
[info] 
[info] Play is run entirely by the community. If you want to keep using it please consider donating:
[info] https://www.playframework.com/sponsors
[info] 
[info]   __              __
[info]   \ \     ____   / /____ _ __  __
[info]    \ \   / __ \ / // __ `// / / /
[info]    / /  / /_/ // // /_/ // /_/ /
[info]   /_/  / .___//_/ \__,_/ \__, /
[info]       /_/               /____/
[info] 
[info] Version 2.8.12+2-f95484fb-SNAPSHOT running Java 11.0.13
[info] 
[info] Play is run entirely by the community. If you want to keep using it please consider donating:
[info] https://www.playframework.com/sponsors
[info] 

--- (Running the application, auto-reloading is enabled) ---

[info] p.c.s.AkkaHttpServer - Listening for HTTP on /127.0.0.1:9000

Server started, use Alt+D to stop

Character detected: -1
Character detected: -1
Character detected: -1
... (1,897 "Character detected: -1" messages removed from the output.)
Character detected: -1
Character detected: -1
Character detected: -1
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$1 (file:/C:/Users/Me/AppData/Local/Coursier/Cache/v1/https/repo1.maven.org/maven2/com/google/inject/guice/4.2.3/guice-4.2.3.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[info] p.a.h.EnabledFilters - Enabled Filters (see <https://www.playframework.com/documentation/latest/Filters>):

    play.filters.csrf.CSRFFilter
    play.filters.headers.SecurityHeadersFilter
    play.filters.hosts.AllowedHostsFilter

[info] play.api.Play - Application started (Dev) (no global state)
```

Once the application has compiled and is running, the `Character detected: -1` messages seemingly no longer appear.

I also tested this with a _Windows_ command script, and it also produces thousands of  `Character detected: -1` messages (2,154 to be precise) before running the application normally.

When running `sbt run` from the command line, it correctly shuts down when Enter is pressed (with a `Character detected: 13` message). Also, when running from within _IntelliJ_, it correctly shuts down when Alt-D is pressed (_without_ an accompanying `Character detected: ` message).

So it looks like the fix is good.

**Comment by mkurz** [comment]
@ndeverge @MikeEngUSA thanks for testing, [I released Play 2.8.13](https://github.com/playframework/playframework/releases/tag/2.8.13) today :+1:

**Comment by MikeEngUSA** [comment]
@mkurz Thanks for the amazingly quick turnaround!

**Comment by mkurz** [comment]
@MikeEngUSA I guess you are on LinkedIn? I send you a message right now.
